### PR TITLE
Alpha dropout

### PR DESCRIPTION
**Pull request by sourabhvarshney111**
This is towards completion of PR #1278.

### CODE REVIEW

**Comment by sourabhvarshney111** [comment]
@rcurtin @zoq 
I have written some comments in #1278. Please tell me about them. Then, I think it is ready for some review.

File: src/mlpack/methods/ann/layer/alpha_dropout.hpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+class AlphaDropout
+{
+ public:
+  /**
+   * Create the Alpha_Dropout object using the specified ratio.
+   *
+   * @param ratio The probability of setting a value to alphaDash.
+   * @param alphaDash The initial value for alphaDash.
+   */
+  AlphaDropout(const double ratio = 0.5, double alphaDash = -alpha * lambda);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/alpha_dropout.hpp]: What about using const ``alphaDash`` and let's use a more descriptive comment for the parameter something like: dropout scaling parameter.

Comment (sourabhvarshney111) [reply src/mlpack/methods/ann/layer/alpha_dropout.hpp]: Done : )

File: src/mlpack/methods/ann/layer/alpha_dropout_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+template<typename InputDataType, typename OutputDataType>
+template<typename Archive>
+void AlphaDropout<InputDataType, OutputDataType>::serialize(
+    Archive& ar,
+    const unsigned int /* version */)
+{
+  ar & BOOST_SERIALIZATION_NVP(ratio);
+  ar & BOOST_SERIALIZATION_NVP(a);
+  ar & BOOST_SERIALIZATION_NVP(b);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/alpha_dropout_impl.hpp]: We should add alphaDash as well.

Comment (sourabhvarshney111) [reply src/mlpack/methods/ann/layer/alpha_dropout_impl.hpp]: Done : )

File: src/mlpack/methods/ann/layer/alpha_dropout.hpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+ * @endcode
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+        typename InputDataType = arma::mat,
+        typename OutputDataType = arma::mat
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/alpha_dropout.hpp]: This is a really minor issue, but if we want to put the typenames on separate lines here, they should be indented only four spaces, not eight. :)

File: src/mlpack/methods/ann/layer/alpha_dropout.hpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+#define MLPACK_METHODS_ANN_LAYER_ALPHA_DROPOUT_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The alpha - dropout layer is a regularizer that randomly with probability 'ratio'
+ * sets input values to alphaDash.
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/alpha_dropout.hpp]: I think that alpha-dropout is meant to be used with the SELU activation function; should we add a note of that in the documentation?  Correct me if I am wrong.

**Comment by zoq** [comment]
Thanks @sourabhvarshney111 @dakshitagrawal97 for the contribution.

**Comment by sourabhvarshney111** [comment]
We will be happy to see you contribute in open source : ) @dakshitagrawal97

**Pull request merged**